### PR TITLE
Allow MFA deletion using challenges

### PR DESCRIPTION
**Pull request by codingllama**
Change DeleteMFADeviceSync so MFA devices may be deleted using a solved authentication challenge directly, instead of using the challenge to create a privilege token first.

This is the first step in replacing the streaming DeleteMFADevice RPC completely with DeleteMFADeviceSync.

#20343

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -3052,17 +3052,39 @@ func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequ
...
 
-	if err := a.verifyUserToken(token, UserTokenTypeRecoveryApproved, UserTokenTypePrivilege); err != nil {
-		return trace.Wrap(err)
+		if err := a.verifyUserToken(token, UserTokenTypeRecoveryApproved, UserTokenTypePrivilege); err != nil {
+			return trace.Wrap(err)
+		}
+
+	case req.ExistingMFAResponse != nil:
+		var err error
+		user, err = authz.GetClientUsername(ctx)
```
Comment (ibeckermayer) [inline lib/auth/auth.go:3071]: nit:
```suggestion
		user, err := authz.GetClientUsername(ctx)
```

File: lib/auth/auth_test.go
Comment (ibeckermayer) [inline lib/auth/auth_test.go:1]: When I run `TestDeleteMFADeviceSync` I'm getting

```
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestDeleteMFADeviceSync$ github.com/gravitational/teleport/lib/auth

--- FAIL: TestDeleteMFADeviceSync (0.52s)
    --- FAIL: TestDeleteMFADeviceSync/Webauthn_using_challenge (0.00s)
        /Users/ibeckermayer/gravitational/teleport/lib/auth/auth_test.go:2257: 
            	Error Trace:	/Users/ibeckermayer/gravitational/teleport/lib/auth/auth_test.go:2257
            	Error:      	Received unexpected error:
            	            	missing parameter user
            	Test:       	TestDeleteMFADeviceSync/Webauthn_using_challenge
            	Messages:   	DeleteMFADeviceSync failed
    --- FAIL: TestDeleteMFADeviceSync/TOTP_using_challenge (0.00s)
        /Users/ibeckermayer/gravitational/teleport/lib/auth/auth_test.go:2257: 
            	Error Trace:	/Users/ibeckermayer/gravitational/teleport/lib/auth/auth_test.go:2257
            	Error:      	Received unexpected error:
            	            	missing parameter user
            	Test:       	TestDeleteMFADeviceSync/TOTP_using_challenge
            	Messages:   	DeleteMFADeviceSync failed
FAIL
FAIL	github.com/gravitational/teleport/lib/auth	1.903s
FAIL
```

File: lib/auth/auth_test.go [hunk 2]
```diff
@@ -2126,56 +2126,134 @@ func TestDeleteMFADeviceSync(t *testing.T) {
...
+	} {
+		testDev, err := RegisterTestDevice(
+			ctx, userClient, spec.name, spec.deviceType, webDev1 /* authenticator */, spec.opts...)
+		require.NoError(t, err, "RegisterTestDevice(%v)", spec.name)
+		allDevs = append(allDevs, testDev)
+	}
+	deleteWeb1 := allDevs[0]
+	deleteWeb2 := allDevs[1]
+	deleteTOTP1 := allDevs[2]
+	deleteTOTP2 := allDevs[3]
```
Comment (ibeckermayer) [inline lib/auth/auth_test.go:2181]:
```suggestion
insertDevice := func(name string, deviceType proto.DeviceType, opts ...TestDeviceOpt) *TestDevice {
		testDev, err := RegisterTestDevice(ctx, userClient, name, deviceType, webDev1 /* authenticator */, opts...)
		require.NoError(t, err, "RegisterTestDevice(%v)", name)
		return testDev
	}
	deleteWeb1 := insertDevice("delete-web1", proto.DeviceType_DEVICE_TYPE_WEBAUTHN)
	deleteWeb2 := insertDevice("delete-web2", proto.DeviceType_DEVICE_TYPE_WEBAUTHN)
	deleteTOTP1 := insertDevice("delete-totp1", proto.DeviceType_DEVICE_TYPE_TOTP, totpOpts...)
	deleteTOTP2 := insertDevice("delete-totp2", proto.DeviceType_DEVICE_TYPE_TOTP, totpOpts...)
	insertDevice("spare-web1", proto.DeviceType_DEVICE_TYPE_WEBAUTHN)
	insertDevice("spare-totp1", proto.DeviceType_DEVICE_TYPE_TOTP, totpOpts...)
```

File: lib/auth/auth_test.go [hunk 3]
```diff
@@ -2126,56 +2126,134 @@ func TestDeleteMFADeviceSync(t *testing.T) {
...
+			deviceType: proto.DeviceType_DEVICE_TYPE_TOTP,
+			opts:       totpOpts,
+		},
+		{
+			name:       "delete-totp2",
+			deviceType: proto.DeviceType_DEVICE_TYPE_TOTP,
+			opts:       totpOpts,
+		},
+		{
+			name:       "spare-web1",
```
Comment (ibeckermayer) [inline lib/auth/auth_test.go:2164]: A comment explaining why "spares" are necessary, to see `TestDeleteMFADeviceSync_lastDevice` for more, could be clarifying.

**Review by ibeckermayer — APPROVED** [review]

File: lib/auth/auth.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/auth.go:3071]: This one is on purpose, we need to user outside of the switch.

File: lib/auth/auth_test.go
Comment (codingllama) [reply lib/auth/auth_test.go:1]: Runs fine here, could it be the `user, err := ...` change you mentioned above?

File: lib/auth/auth_test.go (hunk 2 above)
Comment (codingllama) [reply lib/auth/auth_test.go:2181]: Nice, I've applied a slightly modified snippet.

File: lib/auth/auth_test.go (hunk 3 above)
Comment (codingllama) [reply lib/auth/auth_test.go:2164]: I've changed the second_factor mode to "optional" and dropped the spares, that should make it slightly simpler.

**Comment by codingllama** [comment]
Thanks for the careful review, Isaiah. Comments addressed.

**Review by ravicious — APPROVED** [review]

**Pull request merged**